File: pumper/web/src/eqMath.ts
import { EqBand, EqConfig, FilterType, WidthMode } from "./protocol";

interface Coefficients {
  b0: number;
  b1: number;
  b2: number;
  a1: number;
  a2: number;
}

export interface ResponsePoint {
  frequencyHz: number;
  gainDb: number;
}

const identity: Coefficients = { b0: 1, b1: 0, b2: 0, a1: 0, a2: 0 };

export function buildCoefficients(band: EqBand, sampleRateHz: number): Coefficients {
  if (!band.enabled || Math.abs(band.gainDb) < 0.0001 || band.frequencyHz >= sampleRateHz / 2) return identity;
  const w0 = (2 * Math.PI * band.frequencyHz) / sampleRateHz;
  const sinW0 = Math.sin(w0);
  const cosW0 = Math.cos(w0);
  const a = 10 ** (band.gainDb / 40);
  let alpha: number;
  let b0: number;
  let b1: number;
  let b2: number;
  let a0: number;
  let a1: number;
  let a2: number;

  if (band.type === FilterType.Peaking) {
    alpha =
      band.widthMode === WidthMode.Bandwidth
        ? sinW0 * Math.sinh((Math.log(2) / 2) * band.bandwidthOctaves * (w0 / sinW0))
        : sinW0 / (2 * band.q);
    b0 = 1 + alpha * a;
    b1 = -2 * cosW0;
    b2 = 1 - alpha * a;
    a0 = 1 + alpha / a;
    a1 = -2 * cosW0;
    a2 = 1 - alpha / a;
  } else {
    alpha = (sinW0 / 2) * Math.sqrt((a + 1 / a) * (1 / band.q - 1) + 2);
    const term = 2 * Math.sqrt(a) * alpha;
    if (band.type === FilterType.LowShelf) {
      b0 = a * (a + 1 - (a - 1) * cosW0 + term);
      b1 = 2 * a * (a - 1 - (a + 1) * cosW0);
      b2 = a * (a + 1 - (a - 1) * cosW0 - term);
      a0 = a + 1 + (a - 1) * cosW0 + term;
      a1 = -2 * (a - 1 + (a + 1) * cosW0);
      a2 = a + 1 + (a - 1) * cosW0 - term;
    } else {
      b0 = a * (a + 1 + (a - 1) * cosW0 + term);
      b1 = -2 * a * (a - 1 + (a + 1) * cosW0);
      b2 = a * (a + 1 + (a - 1) * cosW0 - term);
      a0 = a + 1 - (a - 1) * cosW0 + term;
      a1 = 2 * (a - 1 - (a + 1) * cosW0);
      a2 = a + 1 - (a - 1) * cosW0 - term;
    }
  }
  return { b0: b0 / a0, b1: b1 / a0, b2: b2 / a0, a1: a1 / a0, a2: a2 / a0 };
}

function coefficientGainDb(coefficients: Coefficients, frequencyHz: number, sampleRateHz: number): number {
  const w = (2 * Math.PI * frequencyHz) / sampleRateHz;
  const cos1 = Math.cos(w);
  const sin1 = -Math.sin(w);
  const cos2 = Math.cos(2 * w);
  const sin2 = -Math.sin(2 * w);
  const nr = coefficients.b0 + coefficients.b1 * cos1 + coefficients.b2 * cos2;
  const ni = coefficients.b1 * sin1 + coefficients.b2 * sin2;
  const dr = 1 + coefficients.a1 * cos1 + coefficients.a2 * cos2;
  const di = coefficients.a1 * sin1 + coefficients.a2 * sin2;
  return 10 * Math.log10((nr * nr + ni * ni) / (dr * dr + di * di));
}

function buildChain(config: EqConfig, sampleRateHz: number): Coefficients[] {
  return config.bands.map((band) => buildCoefficients(band, sampleRateHz));
}

function chainGainDb(chain: Coefficients[], frequencyHz: number, sampleRateHz: number): number {
  return chain.reduce(
    (gain, coefficients) => gain + coefficientGainDb(coefficients, frequencyHz, sampleRateHz),
    0,
  );
}

export function compositeGainDb(config: EqConfig, sampleRateHz: number, frequencyHz: number): number {
  if (!config.enabled) return 0;
  return config.bands.reduce(
    (gain, band) => gain + coefficientGainDb(buildCoefficients(band, sampleRateHz), frequencyHz, sampleRateHz),
    0,
  );
}

function logFrequency(index: number, count: number): number {
  return 20 * (1000 ** (index / (count - 1)));
}

export function responseCurve(config: EqConfig, sampleRateHz: number, count = 512): ResponsePoint[] {
  const chain = buildChain(config, sampleRateHz);
  return Array.from({ length: count }, (_, index) => {
    const frequencyHz = logFrequency(index, count);
    const filterGain = config.enabled ? chainGainDb(chain, frequencyHz, sampleRateHz) : 0;
    return { frequencyHz, gainDb: config.enabled ? filterGain + config.preampDb : 0 };
  });
}

export function calculateAutoPreamp(config: EqConfig, sampleRateHz: number): { preampDb: number; peakDb: number } {
  if (!config.enabled) return { preampDb: -0.5, peakDb: 0 };
  const chain = buildChain(config, sampleRateHz);
  const count = 4096;
  const frequencies = Array.from({ length: count }, (_, index) => logFrequency(index, count));
  config.bands.filter((band) => band.enabled).forEach((band) => frequencies.push(band.frequencyHz));
  frequencies.sort((a, b) => a - b);

  let peakDb = -Infinity;
  let peakIndex = 0;
  frequencies.forEach((frequency, index) => {
    const gain = chainGainDb(chain, frequency, sampleRateHz);
    if (gain > peakDb) {
      peakDb = gain;
      peakIndex = index;
    }
  });

  const lower = frequencies[Math.max(0, peakIndex - 1)];
  const upper = frequencies[Math.min(frequencies.length - 1, peakIndex + 1)];
  for (let i = 0; i <= 256; i++) {
    const frequency = lower * (upper / lower) ** (i / 256);
    peakDb = Math.max(peakDb, chainGainDb(chain, frequency, sampleRateHz));
  }
  const preampDb = Math.round(Math.min(0, -0.5 - peakDb) * 1000) / 1000;
  return { preampDb, peakDb };
}

File: pumper/android/app/src/main/java/dev/lightwolf/pumper/controller/protocol/EqMath.kt
package dev.lightwolf.pumper.controller.protocol

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.ln
import kotlin.math.log10
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.round
import kotlin.math.sin
import kotlin.math.sinh
import kotlin.math.sqrt

data class ResponsePoint(val frequencyHz: Double, val gainDb: Double)
data class AutoPreampResult(val preampDb: Double, val peakDb: Double)

private data class Coefficients(
    val b0: Double,
    val b1: Double,
    val b2: Double,
    val a1: Double,
    val a2: Double,
)

private val Identity = Coefficients(1.0, 0.0, 0.0, 0.0, 0.0)

object EqMath {
    fun compositeGainDb(config: EqConfig, sampleRateHz: Long, frequencyHz: Double): Double {
        if (!config.enabled) return 0.0
        return config.bands.sumOf { coefficientGainDb(buildCoefficients(it, sampleRateHz), frequencyHz, sampleRateHz) }
    }

    fun responseCurve(config: EqConfig, sampleRateHz: Long, count: Int = 512): List<ResponsePoint> {
        val chain = config.bands.map { buildCoefficients(it, sampleRateHz) }
        return List(count) { index ->
            val frequency = logFrequency(index, count)
            val gain = if (config.enabled) chainGainDb(chain, frequency, sampleRateHz) + config.preampDb else 0.0
            ResponsePoint(frequency, gain)
        }
    }

    fun calculateAutoPreamp(config: EqConfig, sampleRateHz: Long): AutoPreampResult {
        if (!config.enabled) return AutoPreampResult(0.0, 0.0)
        val chain = config.bands.map { buildCoefficients(it, sampleRateHz) }
        val count = 4096
        val frequencies = MutableList(count) { logFrequency(it, count) }
        frequencies += config.bands.filter { it.enabled }.map { it.frequencyHz }
        frequencies.sort()

        var peakDb = Double.NEGATIVE_INFINITY
        var peakIndex = 0
        frequencies.forEachIndexed { index, frequency ->
            val gain = chainGainDb(chain, frequency, sampleRateHz)
            if (gain > peakDb) {
                peakDb = gain
                peakIndex = index
            }
        }

        val lower = frequencies[max(0, peakIndex - 1)]
        val upper = frequencies[min(frequencies.lastIndex, peakIndex + 1)]
        repeat(257) { index ->
            val frequency = lower * (upper / lower).pow(index / 256.0)
            peakDb = max(peakDb, chainGainDb(chain, frequency, sampleRateHz))
        }
        val rounded = round(min(0.0, -peakDb) * 1000.0) / 1000.0
        return AutoPreampResult(if (rounded == -0.0) 0.0 else rounded, peakDb)
    }

    private fun buildCoefficients(band: EqBand, sampleRateHz: Long): Coefficients {
        if (!band.enabled || kotlin.math.abs(band.gainDb) < 0.0001 || band.frequencyHz >= sampleRateHz / 2.0) {
            return Identity
        }
        val w0 = 2.0 * PI * band.frequencyHz / sampleRateHz
        val sinW0 = sin(w0)
        val cosW0 = cos(w0)
        val a = 10.0.pow(band.gainDb / 40.0)
        val alpha: Double
        val b0: Double
        val b1: Double
        val b2: Double
        val a0: Double
        val a1: Double
        val a2: Double

        if (band.type == FilterType.Peaking) {
            alpha = if (band.widthMode == WidthMode.Bandwidth) {
                sinW0 * sinh((ln(2.0) / 2.0) * band.bandwidthOctaves * (w0 / sinW0))
            } else {
                sinW0 / (2.0 * band.q)
            }
            b0 = 1.0 + alpha * a
            b1 = -2.0 * cosW0
            b2 = 1.0 - alpha * a
            a0 = 1.0 + alpha / a
            a1 = -2.0 * cosW0
            a2 = 1.0 - alpha / a
        } else {
            alpha = (sinW0 / 2.0) * sqrt((a + 1.0 / a) * (1.0 / band.q - 1.0) + 2.0)
            val term = 2.0 * sqrt(a) * alpha
            if (band.type == FilterType.LowShelf) {
                b0 = a * (a + 1.0 - (a - 1.0) * cosW0 + term)
                b1 = 2.0 * a * (a - 1.0 - (a + 1.0) * cosW0)
                b2 = a * (a + 1.0 - (a - 1.0) * cosW0 - term)
                a0 = a + 1.0 + (a - 1.0) * cosW0 + term
                a1 = -2.0 * (a - 1.0 + (a + 1.0) * cosW0)
                a2 = a + 1.0 + (a - 1.0) * cosW0 - term
            } else {
                b0 = a * (a + 1.0 + (a - 1.0) * cosW0 + term)
                b1 = -2.0 * a * (a - 1.0 + (a + 1.0) * cosW0)
                b2 = a * (a + 1.0 + (a - 1.0) * cosW0 - term)
                a0 = a + 1.0 - (a - 1.0) * cosW0 + term
                a1 = 2.0 * (a - 1.0 - (a + 1.0) * cosW0)
                a2 = a + 1.0 - (a - 1.0) * cosW0 - term
            }
        }
        return Coefficients(b0 / a0, b1 / a0, b2 / a0, a1 / a0, a2 / a0)
    }

    private fun coefficientGainDb(coefficients: Coefficients, frequencyHz: Double, sampleRateHz: Long): Double {
        val w = 2.0 * PI * frequencyHz / sampleRateHz
        val cos1 = cos(w)
        val sin1 = -sin(w)
        val cos2 = cos(2.0 * w)
        val sin2 = -sin(2.0 * w)
        val nr = coefficients.b0 + coefficients.b1 * cos1 + coefficients.b2 * cos2
        val ni = coefficients.b1 * sin1 + coefficients.b2 * sin2
        val dr = 1.0 + coefficients.a1 * cos1 + coefficients.a2 * cos2
        val di = coefficients.a1 * sin1 + coefficients.a2 * sin2
        return 10.0 * log10((nr * nr + ni * ni) / (dr * dr + di * di))
    }

    private fun chainGainDb(chain: List<Coefficients>, frequencyHz: Double, sampleRateHz: Long): Double =
        chain.sumOf { coefficientGainDb(it, frequencyHz, sampleRateHz) }

    private fun logFrequency(index: Int, count: Int): Double = 20.0 * 1000.0.pow(index.toDouble() / (count - 1))
}

